flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpillingAdaptiveSpanningRecordDeserializer.java

```diff
@@ -597,12 +597,12 @@ private void addNextChunkFromMemorySegment(MemorySegment segment, int offset, in
 				throw new UnsupportedOperationException("Unaligned checkpoint currently do not support spilled " +
 					"records.");
 			} else if (recordLength != -1) {
-				int leftOverSize = leftOverLimit - leftOverStart;
+				int leftOverSize = leftOverData != null ? leftOverLimit - leftOverStart : 0;
```

Comment:
Thanks for finding this bug!

I think the root cause was the state inconsistency among `{leftOverLimit, leftOverStart}` with `leftOverData`. During `#clear()` we only reset `leftOverData` as null, but now reset the derived `{leftOverLimit, leftOverStart}` from `leftOverData`. So we can check the condition only by `leftOverData`. Maybe we can also reset `{leftOverLimit, leftOverStart}` during `#clear()` to keep all consistency.